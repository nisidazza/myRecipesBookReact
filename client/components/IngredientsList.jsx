import React from 'react'
import { apiGetIngredients } from '../apis/recipesApi'

class IngredientsList extends React.Component {
    constructor(props) {
        super(props)

        this.loginData = {
            ingredients: []
        }
        console.log(this.loginData.ingredients)
    }

    componentDidMount() {
        this.fetchIngredients()
    }

    fetchIngredients = () => {
        apiGetIngredients()
            .then(ingredients => {
                this.setState({
                    ingredients: ingredients
                })
            })
    }


    render() {
        return (
            <div>
                <h1>Ingredients List</h1>
                <ul>
                    {this.loginData.ingredients.map(ingredient => {
                        return (
                            <>
                                <li>{ingredient.name}</li>
                            </>
                        )
                    })}
                </ul>
            </div>
        )
    }
}

export default IngredientsList


