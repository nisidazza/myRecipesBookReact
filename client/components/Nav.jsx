import React from "react";
import { Link } from "react-router-dom";
import { logOff, getDecodedToken } from "authenticare/client";
import { IfAuthenticated, IfNotAuthenticated } from "./Authenticated";

class Nav extends React.Component {
  constructor(props) {
    super(props);
  }

  getUserName = () => {
    let myToken = getDecodedToken();
    if (myToken) {
      return myToken.username;
    } else {
      return "";
    }
  };

  handleLogOff = () => {
    logOff();
    this.props.history.push("/");
  };

  render() {
    return (
      <div id="Nav-jsx-component">
        <div
          className="jumbotron-fluid text-center"
          style={{ marginBottom: "0" }}
        >
          <h1 className="display-4 " id="title">
            My Recipe Book
          </h1>
          <h4 id="statement">Keep track of your favourite recipes!</h4>
        </div>
        <nav className="navbar fixed navbar-expand-sm">
          <div className="collapse navbar-collapse" id="navbarSupportedContent">
            <ul className="navbar-nav">
              <li className="nav-item">
                <Link className="nav-link" to="/homepage">
                  Homepage
                </Link>
              </li>
            </ul>
          </div>

          <IfAuthenticated>
            <div
              className="collapse navbar-collapse"
              id="navbarSupportedContent"
            >
              <ul className="navbar-nav ml-auto">
                <li className="nav-item p-2 mt-2">
                  <h5 className="nav-item">Welcome, {this.getUserName()}!</h5>
                </li>
                <li className="nav-item p-0 mt-2">
                  <Link to="#" className="nav-link" onClick={this.handleLogOff}>
                    Log Off
                  </Link>
                </li>
              </ul>
            </div>
          </IfAuthenticated>

          <ul className="navbar-nav ml-auto">
            <IfNotAuthenticated>
              <div className="d-flex p-2">
                <li className="nav-item w-100 pl-2">
                  <Link to="/register" className="nav-link">
                    Register
                  </Link>
                </li>
                <li className="nav-item w-100 pl-2">
                  <Link to="/signin" className="nav-link">
                    Sign In
                  </Link>
                </li>
              </div>
            </IfNotAuthenticated>
          </ul>
        </nav>
        <nav className="navbar sidenav">
          <ul className="navbar-nav">
            <li className="nav-item">
              <Link className="nav-link not-active" to="#">
                Search
              </Link>
            </li>
            <li className="nav-item">
              <Link className="nav-link" to="/listrecipes/">
                Recipes List
              </Link>
            </li>
            <IfAuthenticated>
              <li className="nav-item">
                <Link className="nav-link" to="/addrecipe">
                  Add Recipe
                </Link>
              </li>
            </IfAuthenticated>
            <IfNotAuthenticated></IfNotAuthenticated>
          </ul>
        </nav>
      </div>
    );
  }
}

export default Nav;
