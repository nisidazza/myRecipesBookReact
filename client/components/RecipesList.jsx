import React from 'react'
import { apiGetRecipes, apiDeleteRecipe } from '../apis/recipesApi'
import { Link } from 'react-router-dom'


class RecipesList extends React.Component {
    constructor(props) {
        super(props)

        this.state = {
            recipes: []
        }
    }

    componentDidMount() {
        this.fetchRecipes()
    }

    fetchRecipes = () => {
        apiGetRecipes()
            .then(recipes => {
                this.setState({
                    recipes: recipes
                })
            })
    }
    

    deleteRecipe = (id, i) => {
        apiDeleteRecipe(id)
            .then(hasBeenDeleted => {
                if (hasBeenDeleted) {
                    this.state.recipes.splice(i, 1)
                    this.setState({
                        recipes: this.state.recipes
                    })
                } else {
                    //TODO: 
                }
            })
            .catch(err => this.setState({ error: err.message }))
    }

    viewRecipe = (id) => {
        ap
    }

    render() {
        return (
            <div>
                <h2>Recipes List</h2>
                <ul>
                    {this.state.recipes.map((recipe, i) => {
                        return (
                            <div key={i}>
                                <div>
                                    <li><strong>{recipe.title}</strong></li>
                                    <p>Category: {recipe.category}</p>
                                    {/* <p>{recipe.id}</p> */}
                                    <div>
                                        <button onClick={() => this.deleteRecipe(recipe.id, i)}>Delete</button>
                                        <Link to={`/recipe/${recipe.id}`}>
                                            <button>Update</button>
                                        </Link>
                                        <Link to={`/view/${recipe.id}`}>
                                            <button>View</button>
                                        </Link>
                                    </div>
                                    <hr></hr>
                                </div>
                            </div>
                        )
                    })}
                </ul>
            </div>
        )
    }
}

export default RecipesList