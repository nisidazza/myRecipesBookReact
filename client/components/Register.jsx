import React from "react";
import { register, isAuthenticated } from "authenticare/client";

class Register extends React.Component {
  constructor(props) {
    super(props);

    this.loginData = {
      username: "",
      password: ""
    };
  }

  componentDidMount() {
    let event = new CustomEvent("pageHasChanged", {
      detail: { pageTitle: "Register" }
    });
    document.dispatchEvent(event);
  }

  handleChange = e => {
    this.loginData[e.target.id] = e.target.value;
  };

  handleClick = () => {
    register(this.loginData, {
      baseUrl: process.env.PUBLIC_BASE_API_URL // see .env and webpack.config.js
    }).then(token => {
      if (isAuthenticated()) {
        this.props.history.push("/listrecipes");
      }
    });
  };

  render() {
    return (
      <div id="Register-jsx-component">
        <form className="mx-auto" style={{ maxWidth: "500px", margin: "auto" }}>
          <h4 className="text-center mt-5">Register Form</h4>
          <div className="border mt-4" id="border-shadow">
            <div className="form-group row m-2 mt-4 col-xs-3">
              <div className="input-container mx-auto mt-3 col-xs-3">
                <i className="fa fa-user icon" />
                <input
                  type="text"
                  className="form-control"
                  id="username"
                  placeholder="Username"
                  autoComplete="off"
                  onChange={this.handleChange}
                ></input>
              </div>
            </div>
            <div className="form-group row m-2">
              <div className="input-container mx-auto col-xs-3">
                <i className="fa fa-key icon"></i>
                <input
                  type="password"
                  className="form-control"
                  id="password"
                  placeholder="Password"
                  autoComplete="off"
                  onChange={this.handleChange}
                ></input>
              </div>
            </div>
            <div className="input-container mx-auto mb-4 col-xs-3">
              <button
                className="btn btn-outline-success mx-auto"
                onClick={this.handleClick}
              >
                Register
              </button>
            </div>
          </div>
          {/* TO DO: add confirm password field */}
        </form>
      </div>
    );
  }
}

export default Register;
