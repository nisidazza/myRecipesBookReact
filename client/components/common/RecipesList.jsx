import React from "react";
import RecipeCard from "../pages/recipe/RecipeCard"
import { apiDeleteRecipe } from "../../apis/recipesApi";

class RecipesList extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      recipes: this.props.recipes
    };
  }

  deleteRecipe = (id, i) => {
    apiDeleteRecipe(id)
      .then(hasBeenDeleted => {
        if (hasBeenDeleted) {
          this.state.recipes.splice(i, 1);
          this.setState({
            recipes: this.state.recipes
          });
        } else {
          //TODO:
        }
      })
      .catch(err => this.setState({ error: err.message }));
  };

  render() {
    return (
      <div id="RecipesList-jsx-component">
        <div className="row">
          {this.state.recipes.map((recipe, i) => {
            return (
              <RecipeCard key={i} openInNewTab={true} recipe={recipe} onClickDelete={() => this.deleteRecipe(recipe.id, i)}/>
            );
          })}
        </div>
      </div>
    );
  }
}

export default RecipesList;
