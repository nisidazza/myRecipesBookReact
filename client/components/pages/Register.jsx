import React from "react";
import { register, isAuthenticated } from "authenticare/client";
import Validator from "../common/utilities/Validator";

class Register extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      loginData: {
        username: "",
        password: "",
      },
      confirmPassword: "",
      errorMessage: [],
      isValid: true
    };

    this.validator = React.createRef();
  }

  componentDidMount() {
    let event = new CustomEvent("pageHasChanged", {
      detail: { pageTitle: "Register" }
    });
    document.dispatchEvent(event);
  }

  handleLoginData = e => {
    this.state.loginData[e.target.id] = e.target.value;
  };

  handlePasswordConfirmation = e => {
    this.setState({
      confirmPassword: e.target.value
    });
  };

  handleClick = e => {
    e.preventDefault();
    if (this.validator.current.validate()) {
      register(this.state.loginData, {
        baseUrl: process.env.PUBLIC_BASE_API_URL // see .env and webpack.config.js
      })
        .then(response => {
          if (isAuthenticated()) {
            this.props.history.push("/listrecipes");
          }
        })
        .catch(error => {
          if (
            error &&
            error.response &&
            error.response.body &&
            error.response.body.errorType == "USERNAME_UNAVAILABLE"
          ) {
            this.validator.current.showError("Sorry, that email address already exists!");
          } else {
            this.validator.current.showError(
              "Something went wrong. Please,try again!"
            );
            throw error;
          }
        });
    }
  };

  getValidationRules = () => {
    let rules = [];

    rules.push({
      conditional: () => {
        return (
          this.state.loginData.username !== "" &&
          this.state.loginData.username.match(/\S+@\S+\.\S+/)
        );
      },
      errorMessage: "Please, insert a valid email address"
    });

    rules.push({
      conditional: () => {
        return (
          this.state.loginData.password.match(/[a-z]/g) &&
          this.state.loginData.password.match(/[A-Z]/g) &&
          this.state.loginData.password.match(/[0-9]/g) &&
          this.state.loginData.password.match(/[^a-zA-Z\d]/g) &&
          this.state.loginData.password.length >= 8
        );
      },
      errorMessage:
        "A valid password must contain at least 1 uppercase character, at least 1 lowercase character, at least 1 digit, at least 1 special character and minimum 8 characters"
    });

    rules.push({
      conditional: () => {
        return this.state.confirmPassword == this.state.loginData.password;
      },
      errorMessage: "Password and Confirm Password do not match"
    });

    return rules;
  };

  render() {
    return (
      <div id="Register-jsx-component">
        <form className="mx-auto" style={{ maxWidth: "500px", margin: "auto" }}>
          <h4 className="text-center mt-5">Register Form</h4>
          <div className="border mt-4" id="border-shadow">
            <div className="form-group row m-2 mt-4">
              <div className="input-container mx-auto mt-3 col-md-9">
                <i className="fa fa-user icon" />
                <input
                  type="text"
                  className="form-control"
                  id="username"
                  placeholder="Email address"
                  autoComplete="off"
                  onChange={this.handleLoginData}
                ></input>
              </div>
            </div>
            <div className="form-group row m-2">
              <div className="input-container mx-auto col-md-9">
                <i className="fa fa-key icon"></i>
                <input
                  type="password"
                  className="form-control"
                  id="password"
                  placeholder="Password"
                  autoComplete="off"
                  onChange={this.handleLoginData}
                ></input>
              </div>
            </div>
            <div className="form-group row m-2">
              <div className="input-container mx-auto col-md-9">
                <i className="fa fa-key icon"></i>
                <input
                  type="password"
                  className="form-control"
                  id="confirm-password"
                  placeholder="Confirm Password"
                  autoComplete="off"
                  onChange={this.handlePasswordConfirmation}
                ></input>
              </div>
            </div>
            <div className="input-container mx-auto mb-4 col-md-9">
              <input
                value="Register"
                type="submit"
                className="btn btn-outline-success mx-auto"
                onClick={this.handleClick}
              ></input>
            </div>
            <div>
              <Validator
                rules={this.getValidationRules()}
                ref={this.validator}
              />
            </div>
          </div>
        </form>
      </div>
    );
  }
}

export default Register;
