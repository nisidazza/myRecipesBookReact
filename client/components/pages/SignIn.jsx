import React from "react";
import { signIn, isAuthenticated } from "authenticare/client";
import Validator from "../common/utilities/Validator";

class SignIn extends React.Component {
  constructor(props) {
    super(props);

    this.loginData = {
      username: "",
      password: ""
    };

    this.validator = React.createRef();
  }

  componentDidMount() {
    let event = new CustomEvent("pageHasChanged", {
      detail: { pageTitle: "Sign-In" }
    });
    document.dispatchEvent(event);
  }

  handleChange = e => {
    this.loginData[e.target.id] = e.target.value;
  };

  handleSubmit = e => {
    e.preventDefault();
    signIn(this.loginData, {
      baseUrl: process.env.PUBLIC_BASE_API_URL // see .env and webpack.config.js
    })
      .then(token => {
        if (isAuthenticated()) {
          this.props.history.push("/listrecipes");
        }
      })
      .catch(error => {
        if (
          error &&
          error.response &&
          error.response.body &&
          error.response.body.errorType == "INVALID_CREDENTIALS"
        ) {
          this.validator.current.showError(
            "Sorry, wrong username + password combination!"
          );
        } else {
          this.validator.current.showError(
            "Something went wrong. Please,try again!"
          );
          throw error;
        }
      });
  };

  render() {
    return (
      <div id="SignIn-jsx-component">
        <form
          className="mx-auto"
          style={{ maxWidth: "500px", margin: "auto" }}
          onSubmit={this.handleSubmit}
        >
          <h4 className="text-center mt-5">Sign-In Form </h4>
          <div className="border mt-4" id="border-shadow">
            <div className="form-group row m-2 mt-4">
              <div className="input-container mx-auto mt-3 col-xs-4">
                <i className="fa fa-user icon" />
                <input
                  type="text"
                  className="form-control"
                  id="username"
                  placeholder="Email address"
                  autoComplete="off"
                  onChange={this.handleChange}
                ></input>
              </div>
            </div>
            <div className="form-group row m-2">
              <div className="input-container mx-auto col-xs-4">
                <i className="fa fa-key icon"></i>
                <input
                  type="password"
                  className="form-control"
                  id="password"
                  placeholder="Password"
                  autoComplete="off"
                  onChange={this.handleChange}
                ></input>
              </div>
            </div>
            <div className="input-container mx-auto mb-3 col-xs-3">
              <input
                value="Sign In"
                type="submit"
                className="btn btn-outline-success mx-auto"
              />
            </div>
            <div>
              <Validator ref={this.validator} />
            </div>
            <div className="forgot-psw">
              <a href="#/recovery/email/" target="_blank">Forgot your Password?</a>
            </div>
          </div>
        </form>
      </div>
    );
  }
}

export default SignIn;
