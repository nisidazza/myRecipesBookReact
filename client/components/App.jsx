import React from 'react'
import { HashRouter as Router, Route, Link } from 'react-router-dom'
import IngredientsList from './IngredientsList'
import RecipesList from './RecipesList'
import Nav from './Nav'
import Register from './Register'
import SignIn from './SignIn'


class App extends React.Component {
    render() {
        return (
            <Router>
                <h1>My Recipes Book</h1>
                <Route path='/' component={Nav} />
                <Route path='/register' component={Register} />
                <Route path='/signin' component={SignIn} />
                <Route path='/ingredients' component={IngredientsList} />
                <Route path='/recipes' component={RecipesList} />
            </Router>
        )
    }
}

export default App