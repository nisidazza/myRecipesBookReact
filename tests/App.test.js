import React from 'react'
import renderer from 'react-test-renderer'
import {shallow} from 'enzyme'
import App from '../client/components/App'
import SignIn from '../client/components/SignIn'
import Register from '../client/components/Register'
import IngredientsList from '../client/components/IngredientsList'
import RecipesList from '../client/components/RecipesList'
import Nav from '../client/components/Nav'
import {Route} from 'react-router-dom'


describe('My Test Suite', () => {
    it('My Test Case', () => {
        expect(true).toEqual(true)
    }),
    it('My Test Case2', () => {
        expect(true).toEqual(true) 
    }),
    it('renders correctly', () => {
        const component = renderer.create(<App />)
        let tree = component.toJSON()
        expect(tree).toMatchSnapshot()
    })
})

let pathMap = {}
describe('routes using array of routers', () => {
    beforeAll(() => {
        const wrapper = shallow(<App />);        
        console.log("wrapper: ")
        console.log(wrapper)
        pathMap = wrapper.find(Route).reduce((pathMap, route) => {
            const routeProps = route.props() 
            pathMap[routeProps.path] = routeProps.component
            return pathMap
        }, {})
        console.log("pathMap: ")
        console.log(pathMap)
    })
    it('should show Nav component for / router', () => {
        expect(pathMap['/']).toBe(Nav)
    }),
    it('should show Register component for /register', () => {
        expect(pathMap['/register']).toBe(Register)
    }),
    it('should show SignIn component for /register', () => {
        expect(pathMap['/signin']).toBe(SignIn)
    }),
    it('should show IngredientsList component for /ingredients', () => {
        expect(pathMap['/ingredients']).toBe(IngredientsList)
    }),
    it('should show RecipesList component for /recipes', () => {
        expect(pathMap['/recipes']).toBe(RecipesList)
    })
})

