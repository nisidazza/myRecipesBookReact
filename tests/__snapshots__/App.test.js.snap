// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`My Test Suite renders correctly 1`] = `
Array [
  <h1>
    My Recipes Book
  </h1>,
  <a
    href="#/"
    onClick={[Function]}
  >
    Home
  </a>,
  <a
    href="#/register"
    onClick={[Function]}
  >
    Register
  </a>,
  <a
    href="#/signin"
    onClick={[Function]}
  >
    Sign In
  </a>,
]
`;
